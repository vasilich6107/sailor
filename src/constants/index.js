export const APPLICANT_STATES = {
  APPLIED: 'APPLIED',
  INTERVIEWING: 'INTERVIEWING',
  HIRED: 'HIRED',
};
