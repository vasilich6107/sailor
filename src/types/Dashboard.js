// @flow

export type DashboardFilterType = {|
  name: string,
  city: string,
|};

export type DashboardColumnsType = Array<string>;
