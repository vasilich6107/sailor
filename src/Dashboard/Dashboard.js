// @flow
import React, { PureComponent, Fragment } from 'react';
import { APPLICANT_STATES } from '../constants';

import type { ApplicantType, ApplicantDataType } from '../types/Applicant';
import type { DashboardFilterType } from '../types/Dashboard';
import DashboardTable from './DashboardTable';
import DashboardFilter from './DashboardFilter';

export const DASHBOARD_COLUMNS: Array<string> = [
  APPLICANT_STATES.APPLIED,
  APPLICANT_STATES.INTERVIEWING,
  APPLICANT_STATES.HIRED,
];

const filterIndexOf = (subject: string, needle: string): boolean => (needle ? subject.indexOf(needle) !== -1 : true);

type Props = {};

type State = {
  filter: DashboardFilterType,
  applicants: ApplicantDataType,
};

class Dashboard extends PureComponent<Props, State> {
  state = {
    filter: {
      name: localStorage.getItem('nameFilter') || '',
      city: localStorage.getItem('cityFilter') || '',
    },
    applicants: {
      loading: false,
      data: {},
      error: null,
    },
  };

  componentDidMount(): void {
    this.setState(
      previousState => ({
        applicants: {
          ...previousState.applicants,
          loading: true,
        },
      }),
      this.fetchData
    );
  }

  fetchData = () =>
    fetch('https://randomuser.me/api/?nat=gb&results=5&seed=56308e5db1a68b03')
      .then(response => response.json())
      .then(({ results }) => {
        const data = results.reduce((acc, applicant) => {
          acc[applicant.login.uuid] = {
            ...applicant,
            state: APPLICANT_STATES.APPLIED,
          };

          return acc;
        }, {});

        this.setState(previousState => ({
          applicants: {
            ...previousState.applicants,
            loading: false,
            data,
          },
        }));
      })
      .catch(error => {
        this.setState(previousState => ({
          applicants: {
            ...previousState.applicants,
            loading: false,
            error,
          },
        }));
      });

  isVisible = (applicant: ApplicantType): boolean => {
    const { filter } = this.state;

    return (
      (filterIndexOf(applicant.name.first, filter.name) || filterIndexOf(applicant.name.last, filter.name)) &&
      filterIndexOf(applicant.location.city, this.state.filter.city)
    );
  };

  getDashboardApplicants = () => {
    const applicants: Array<ApplicantType> = (Object.values(this.state.applicants.data): Array<any>);

    return applicants.reduce(
      (acc, applicant: ApplicantType) => {
        if (this.isVisible(applicant)) {
          acc[applicant.state].push(applicant);
        }

        return acc;
      },
      {
        [APPLICANT_STATES.APPLIED]: [],
        [APPLICANT_STATES.INTERVIEWING]: [],
        [APPLICANT_STATES.HIRED]: [],
      }
    );
  };

  handleApplicantStateChange = (newState: string) => (uuid: string) => {
    this.setState(({ applicants }) => {
      const data = {
        ...applicants.data,
        [uuid]: {
          ...applicants.data[uuid],
          state: newState,
        },
      };

      return {
        applicants: {
          ...applicants,
          data,
        },
      };
    });
  };

  handleFilterChange = (name: string) => (e: SyntheticInputEvent<HTMLInputElement>): void => {
    const {
      target: { value },
    } = e;
    localStorage.setItem(`${name}Filter`, value);

    this.setState(prevState => ({
      filter: {
        ...prevState.filter,
        [name]: value,
      },
    }));
  };

  render() {
    const { applicants, filter } = this.state;

    if (applicants.loading) {
      return <div>Loading</div>;
    }

    if (applicants.error) {
      return <div>{applicants.error.message}</div>;
    }

    return (
      <Fragment>
        <DashboardFilter initialValues={filter} onFilterChange={this.handleFilterChange} />
        <DashboardTable
          columnOrder={DASHBOARD_COLUMNS}
          applicants={this.getDashboardApplicants()}
          onStateChange={this.handleApplicantStateChange}
        />
      </Fragment>
    );
  }
}

export default Dashboard;
