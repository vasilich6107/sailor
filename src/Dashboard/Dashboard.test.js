import React from 'react';
import { shallow } from 'enzyme';
import Dashboard from './Dashboard';
import { APPLICANT_STATES } from '../constants';

const sleep = ms => new Promise(resolve => setTimeout(resolve, ms));

jest.mock('./DashboardColumn', () => 'DashboardColumn');

const applicants = {
  results: [
    {
      location: {
        city: 'bath',
      },
      name: {
        first: 'francis',
        last: 'stewart',
        title: 'mr',
      },
      login: {
        uuid: '5d204000-387b-4741-8ba9-7c121ae6adab',
      },
      picture: {
        large: 'https://randomuser.me/api/portraits/men/15.jpg',
        medium: 'https://randomuser.me/api/portraits/med/men/15.jpg',
        thumbnail: 'https://randomuser.me/api/portraits/thumb/men/15.jpg',
      },
    },
    {
      location: {
        city: 'nottingham',
      },
      name: {
        first: 'harvey',
        last: 'clarke',
        title: 'mr',
      },
      login: {
        uuid: '40fdd877-b7f1-47bd-8c6b-8b2129bbf3fd',
      },
      picture: {
        large: 'https://randomuser.me/api/portraits/men/15.jpg',
        medium: 'https://randomuser.me/api/portraits/med/men/15.jpg',
        thumbnail: 'https://randomuser.me/api/portraits/thumb/men/15.jpg',
      },
    },
    {
      location: {
        city: 'coventry',
      },
      name: {
        first: 'mia',
        last: 'white',
        title: 'mr',
      },
      login: {
        uuid: '674ed62e-a5f2-42a1-9ba8-2c00c3671801',
      },
      picture: {
        large: 'https://randomuser.me/api/portraits/men/15.jpg',
        medium: 'https://randomuser.me/api/portraits/med/men/15.jpg',
        thumbnail: 'https://randomuser.me/api/portraits/thumb/men/15.jpg',
      },
    },
  ],
};

let storeMock = {};
const localStorageMock = (() => ({
  getItem(key) {
    return storeMock[key] || null;
  },
  setItem(key, value) {
    storeMock[key] = value.toString();
  },
  removeItem(key) {
    delete storeMock[key];
  },
  clear() {
    storeMock = {};
  },
}))();

const fetchSuccess = jest.fn().mockImplementation(
  () =>
    new Promise(resolve => {
      resolve({
        json: () => applicants,
      });
    })
);

const fetchError = jest.fn().mockImplementation(
  () =>
    new Promise((resolve, reject) => {
      reject(new Error('Network error'));
    })
);

Object.defineProperty(window, 'localStorage', {
  value: localStorageMock,
});

it('Renders dashboard in Loading state', async () => {
  const component = shallow(<Dashboard />);
  expect(component).toMatchSnapshot();
});

it('Renders dashboard with error', async () => {
  global.fetch = fetchError;
  const component = shallow(<Dashboard />);
  await sleep(100);

  expect(component).toMatchSnapshot();
});

it('Renders dashboard with loaded data', async () => {
  global.fetch = fetchSuccess;

  const handleApplicantStateChangeMock = jest.fn();

  const component = shallow(<Dashboard />);
  component.instance().handleApplicantStateChange = handleApplicantStateChangeMock;
  await sleep(100);

  expect(component).toMatchSnapshot();

  expect(handleApplicantStateChangeMock.mock.calls.length).toBe(4);

  expect(handleApplicantStateChangeMock.mock.calls[0][0]).toBe(APPLICANT_STATES.INTERVIEWING);
  expect(handleApplicantStateChangeMock.mock.calls[1][0]).toBe(APPLICANT_STATES.HIRED);
  expect(handleApplicantStateChangeMock.mock.calls[2][0]).toBe(APPLICANT_STATES.APPLIED);
  expect(handleApplicantStateChangeMock.mock.calls[3][0]).toBe(APPLICANT_STATES.INTERVIEWING);
});

it('Renders dashboard and move applicant between columns', async () => {
  global.fetch = fetchSuccess;

  const component = shallow(<Dashboard />);
  await sleep(100);

  component.instance().handleApplicantStateChange(APPLICANT_STATES.INTERVIEWING)(applicants.results[0].login.uuid);
  expect(component).toMatchSnapshot();

  component.instance().handleApplicantStateChange(APPLICANT_STATES.HIRED)(applicants.results[0].login.uuid);
  expect(component).toMatchSnapshot();

  component.instance().handleApplicantStateChange(APPLICANT_STATES.APPLIED)(applicants.results[0].login.uuid);
  expect(component).toMatchSnapshot();
});

describe('Renders dashboard and filters applicants', () => {
  let component;

  beforeAll(async () => {
    global.fetch = fetchSuccess;
    component = shallow(<Dashboard />);

    await sleep(100);
  });

  it('Filter by name and country', () => {
    const nameValue = 'fr';
    const cityValue = 'ba';

    component.find('#nameInput').simulate('change', { target: { value: nameValue } });
    expect(component).toMatchSnapshot();

    component.find('#cityInput').simulate('change', { target: { value: cityValue } });
    expect(component).toMatchSnapshot();

    expect(storeMock.nameFilter).toBe(nameValue);
    expect(storeMock.cityFilter).toBe(cityValue);
  });

  it('Resets filters', async () => {
    const nameValue = '';
    const cityValue = '';

    component.find('#nameInput').simulate('change', { target: { value: nameValue } });
    expect(component).toMatchSnapshot();

    component.find('#cityInput').simulate('change', { target: { value: cityValue } });
    expect(component).toMatchSnapshot();

    expect(storeMock.nameFilter).toBe(nameValue);
    expect(storeMock.cityFilter).toBe(cityValue);
  });
});
