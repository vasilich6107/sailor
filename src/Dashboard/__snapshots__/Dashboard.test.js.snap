// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders dashboard and move applicant between columns 1`] = `
<Fragment>
  <DashboardFilter
    initialValues={
      Object {
        "city": "",
        "name": "",
      }
    }
    onFilterChange={[Function]}
  />
  <DashboardTable
    applicants={
      Object {
        "APPLIED": Array [
          Object {
            "location": Object {
              "city": "nottingham",
            },
            "login": Object {
              "uuid": "40fdd877-b7f1-47bd-8c6b-8b2129bbf3fd",
            },
            "name": Object {
              "first": "harvey",
              "last": "clarke",
              "title": "mr",
            },
            "picture": Object {
              "large": "https://randomuser.me/api/portraits/men/15.jpg",
              "medium": "https://randomuser.me/api/portraits/med/men/15.jpg",
              "thumbnail": "https://randomuser.me/api/portraits/thumb/men/15.jpg",
            },
            "state": "APPLIED",
          },
          Object {
            "location": Object {
              "city": "coventry",
            },
            "login": Object {
              "uuid": "674ed62e-a5f2-42a1-9ba8-2c00c3671801",
            },
            "name": Object {
              "first": "mia",
              "last": "white",
              "title": "mr",
            },
            "picture": Object {
              "large": "https://randomuser.me/api/portraits/men/15.jpg",
              "medium": "https://randomuser.me/api/portraits/med/men/15.jpg",
              "thumbnail": "https://randomuser.me/api/portraits/thumb/men/15.jpg",
            },
            "state": "APPLIED",
          },
        ],
        "HIRED": Array [],
        "INTERVIEWING": Array [
          Object {
            "location": Object {
              "city": "bath",
            },
            "login": Object {
              "uuid": "5d204000-387b-4741-8ba9-7c121ae6adab",
            },
            "name": Object {
              "first": "francis",
              "last": "stewart",
              "title": "mr",
            },
            "picture": Object {
              "large": "https://randomuser.me/api/portraits/men/15.jpg",
              "medium": "https://randomuser.me/api/portraits/med/men/15.jpg",
              "thumbnail": "https://randomuser.me/api/portraits/thumb/men/15.jpg",
            },
            "state": "INTERVIEWING",
          },
        ],
      }
    }
    columnOrder={
      Array [
        "APPLIED",
        "INTERVIEWING",
        "HIRED",
      ]
    }
    onStateChange={[Function]}
  />
</Fragment>
`;

exports[`Renders dashboard and move applicant between columns 2`] = `
<Fragment>
  <DashboardFilter
    initialValues={
      Object {
        "city": "",
        "name": "",
      }
    }
    onFilterChange={[Function]}
  />
  <DashboardTable
    applicants={
      Object {
        "APPLIED": Array [
          Object {
            "location": Object {
              "city": "nottingham",
            },
            "login": Object {
              "uuid": "40fdd877-b7f1-47bd-8c6b-8b2129bbf3fd",
            },
            "name": Object {
              "first": "harvey",
              "last": "clarke",
              "title": "mr",
            },
            "picture": Object {
              "large": "https://randomuser.me/api/portraits/men/15.jpg",
              "medium": "https://randomuser.me/api/portraits/med/men/15.jpg",
              "thumbnail": "https://randomuser.me/api/portraits/thumb/men/15.jpg",
            },
            "state": "APPLIED",
          },
          Object {
            "location": Object {
              "city": "coventry",
            },
            "login": Object {
              "uuid": "674ed62e-a5f2-42a1-9ba8-2c00c3671801",
            },
            "name": Object {
              "first": "mia",
              "last": "white",
              "title": "mr",
            },
            "picture": Object {
              "large": "https://randomuser.me/api/portraits/men/15.jpg",
              "medium": "https://randomuser.me/api/portraits/med/men/15.jpg",
              "thumbnail": "https://randomuser.me/api/portraits/thumb/men/15.jpg",
            },
            "state": "APPLIED",
          },
        ],
        "HIRED": Array [
          Object {
            "location": Object {
              "city": "bath",
            },
            "login": Object {
              "uuid": "5d204000-387b-4741-8ba9-7c121ae6adab",
            },
            "name": Object {
              "first": "francis",
              "last": "stewart",
              "title": "mr",
            },
            "picture": Object {
              "large": "https://randomuser.me/api/portraits/men/15.jpg",
              "medium": "https://randomuser.me/api/portraits/med/men/15.jpg",
              "thumbnail": "https://randomuser.me/api/portraits/thumb/men/15.jpg",
            },
            "state": "HIRED",
          },
        ],
        "INTERVIEWING": Array [],
      }
    }
    columnOrder={
      Array [
        "APPLIED",
        "INTERVIEWING",
        "HIRED",
      ]
    }
    onStateChange={[Function]}
  />
</Fragment>
`;

exports[`Renders dashboard and move applicant between columns 3`] = `
<Fragment>
  <DashboardFilter
    initialValues={
      Object {
        "city": "",
        "name": "",
      }
    }
    onFilterChange={[Function]}
  />
  <DashboardTable
    applicants={
      Object {
        "APPLIED": Array [
          Object {
            "location": Object {
              "city": "bath",
            },
            "login": Object {
              "uuid": "5d204000-387b-4741-8ba9-7c121ae6adab",
            },
            "name": Object {
              "first": "francis",
              "last": "stewart",
              "title": "mr",
            },
            "picture": Object {
              "large": "https://randomuser.me/api/portraits/men/15.jpg",
              "medium": "https://randomuser.me/api/portraits/med/men/15.jpg",
              "thumbnail": "https://randomuser.me/api/portraits/thumb/men/15.jpg",
            },
            "state": "APPLIED",
          },
          Object {
            "location": Object {
              "city": "nottingham",
            },
            "login": Object {
              "uuid": "40fdd877-b7f1-47bd-8c6b-8b2129bbf3fd",
            },
            "name": Object {
              "first": "harvey",
              "last": "clarke",
              "title": "mr",
            },
            "picture": Object {
              "large": "https://randomuser.me/api/portraits/men/15.jpg",
              "medium": "https://randomuser.me/api/portraits/med/men/15.jpg",
              "thumbnail": "https://randomuser.me/api/portraits/thumb/men/15.jpg",
            },
            "state": "APPLIED",
          },
          Object {
            "location": Object {
              "city": "coventry",
            },
            "login": Object {
              "uuid": "674ed62e-a5f2-42a1-9ba8-2c00c3671801",
            },
            "name": Object {
              "first": "mia",
              "last": "white",
              "title": "mr",
            },
            "picture": Object {
              "large": "https://randomuser.me/api/portraits/men/15.jpg",
              "medium": "https://randomuser.me/api/portraits/med/men/15.jpg",
              "thumbnail": "https://randomuser.me/api/portraits/thumb/men/15.jpg",
            },
            "state": "APPLIED",
          },
        ],
        "HIRED": Array [],
        "INTERVIEWING": Array [],
      }
    }
    columnOrder={
      Array [
        "APPLIED",
        "INTERVIEWING",
        "HIRED",
      ]
    }
    onStateChange={[Function]}
  />
</Fragment>
`;

exports[`Renders dashboard in Loading state 1`] = `
<div>
  Loading
</div>
`;

exports[`Renders dashboard with error 1`] = `
<div>
  Network error
</div>
`;
